OpenRA.Mods.Cnc/Traits/Render/WithReloadingTurret.cs

```diff
@@ -0,0 +1,81 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Renders ammo-dependent turret graphics for units with the Turreted trait.")]
+	public class WithReloadingTurretInfo : WithTurretInfo, Requires<AmmoPoolInfo>, Requires<ArmamentInfo>
+	{
+		[Desc("AmmoPool to use for ammo-dependent sequences.")]
+		public readonly string AmmoPoolName = null;
+
+		[Desc("How many reload stages does this turret have. Defaults to AmmoPool's Ammo.",
+			"Adds current reload stage to Sequence as suffix when a matching AmmoPool is present.")]
+		public readonly int ReloadStages = -1;
+
+		public override object Create(ActorInitializer init) { return new WithReloadingTurret(init.Self, this); }
+	}
+
+	public class WithReloadingTurret : WithTurret
+	{
+		readonly int reloadStages;
+		readonly AmmoPool ammoPool;
+		string sequence;
+		string ammoSuffix;
+
+		public WithReloadingTurret(Actor self, WithReloadingTurretInfo info)
+			: base(self, info)
+		{
+			ammoPool = self.TraitsImplementing<AmmoPool>()
+				.First(a => a.Info.Name == info.AmmoPoolName);
```

Comment:
Using `FirstOrDefault` and then throwing your own exception if it returns null will be more modder-friendly than the exception that `First` throws by itself.